### DISCUSSION TITLE
Tomo in production

### DISCUSSION THREAD

**Discussion by francois0690** [mattbrictson/tomo#286] [Q&A] [answered]
Hi,

I'm looking for best practice for using Tomo in a production environnement.

- Size of `/var/www/my-app/releases/` folder

How do I use Tomo to delete all other folders except the currently used version?
(I can do it manually but I can do a mistake with rm command...)

- Multiple update for load-balancing

I have two front end Rails server for load balancing, I would like to push a new version to both at the same time.
Maybe can I open two terminals and run the following two commands at the same time?
```
tomo deploy --config .tomo/config_host1.rb
tomo deploy --config .tomo/config_host2.rb
```

- New release but same DB 

I have another serveur for Posgresql only.
If I want to keep my data could you please confirm me that the best way to do that is to disable the `run "rails:db_seed"` task in the `deploy do` block in `config.rb` file?
This will not prevent eventual migrations but will not destroy the data.

All other advice to secure the process of updating a production application is welcome.

Thanks for Tomo !

**Comment by mattbrictson** [comment, marked as answer]
Hi @francois0690 ! Good questions.

> How do I use Tomo to delete all other folders except the currently used version?

This happens automatically when the [core:clean_releases](https://tomo-deploy.com/plugins/core/#coreclean_releases) task runs (it should be near the end of your `deploy` block). By default, it will keep the most recent 10 release folders and delete the older ones. If you want to keep only 1 release, add this setting to your `.tomo/config.rb` file:

```ruby
set keep_releases: 1
```

> Multiple update for load-balancing

Tomo is designed to deploy to multiple servers simultaneously. In your `.tomo/config.rb`, add a [host](https://tomo-deploy.com/configuration/#hostaddress-options) line for each server, like this:

```ruby
host "deployer@www1.example.com"
host "deployer@www2.example.com"
```

Now you can run `tomo setup` to set up both machines simultaneously, and `tomo deploy` to deploy to both simultaneously.

In practice, you will also want to use [roles](https://tomo-deploy.com/configuration/#rolename-runs) to avoid doubling-up on certain operations. For example, you probably only want database tasks like `rails:db_migrate` to run one on server, not both. One way to do that is to give one host a "primary" role (you can name the role whatever you want), and then tell tomo to only run those database tasks on the primary host. It would look like this:

```ruby
host "deployer@www1.example.com", roles: ["primary"]
host "deployer@www2.example.com"

role :primary, runs: ["rails:db_*", "rails:console", "rails:log_tail"]
```

> If I want to keep my data could you please confirm me that the best way to do that is to disable the run "rails:db_seed" task

Yes, there is no harm in removing the `run "rails:db_seed"` task from the `deploy` block. I would suggest you keep it in the `setup` block. That way, you can use `tomo setup` to initialize the database once, prior to your first deploy (the seeds will run during setup). After that, you can use `tomo deploy` multiple times and the seeds will not run if you remove that task from the `deploy` block.

> All other advice to secure the process of updating a production application is welcome

As a general practice, I suggest writing your seeds so that they are [idempotent](https://www.cookieshq.co.uk/posts/idempotent-seed-files). In my opinion, running `rails db:seed` should not delete or modify data. That is why the default config in tomo assumes that it is safe to run the seeds on every deploy.

**Reply by francois0690** [reply]
Great! Thank you for your answers! :+1: :pray: :clap:
